Нагдалян Арсен. Homework 2

ArsenHD opened the pull request.

ArsenHD committed:
Solution

ArsenHD committed:
Update architecture

ArsenHD committed:
Add some tests

ArsenHD committed:
Move variable substitution to a separate class

ArsenHD committed:
Add more tests

ArsenHD committed:
Add CI script

ArsenHD committed:
Update README.md

ArsenHD committed:
Rework API in order to support infinite input

ArsenHD committed:
Update tests according to the new API

ArsenHD committed:
Remove explicit input specification for commands, because now it is read from an input stream + unite arguments and options

ArsenHD committed:
Support grep command + improve options parsing

ArsenHD committed:
Fix pipeline command connection

ArsenHD committed:
Add tests for grep command

ArsenHD committed:
Add new branch trigger to CI

ArsenHD committed:
Update README.md

ArsenHD committed:
Do not print extra newlines in cat command

FirstTimeInForever commented on README.md:
```diff
@@ -1 +1,15 @@
+  * echo
+  * pwd
+  * wc
+  * grep
```
Опять же, стоит написать синтаксис команды и вообще поправить `README.md`.

ArsenHD replied:
Обновил README, сейчас он более подробный

FirstTimeInForever commented on src/ru/itmo/sd/shell/cli/command/GrepCommand.kt:
```diff
@@ -0,0 +1,96 @@
+    private fun processFile() = execution {
+        val file = File(fileName!!)
+        if (!file.isFile) {
+            writeLine("grep: ${file.name}: No such file or directory")
```
Логично было бы писать это в `output`.

ArsenHD replied:
Это сообщение сейчас будет выводиться в `output` стрим, переданный команде.

FirstTimeInForever commented on test/ru/itmo/sd/shell/cli/command/GrepCommandTest.kt:
```diff
@@ -0,0 +1,58 @@
+import org.junit.jupiter.params.provider.MethodSource
+import ru.itmo.sd.shell.cli.util.red
+
+class GrepCommandTest : AbstractSimpleCommandTest() {
```
Нужно добавить тест на несуществующий файл и невалидные ключи. `-1`.

ArsenHD replied:
[Done](https://github.com/ArsenHD/SD-Homework1/pull/2/commits/d1309aae49f1d9cf538def6ada178ac4ca344532)

FirstTimeInForever reviewed (changes requested):
По функционалу:
* Падает на:
```
grep cat README.md
Exception in thread "main" java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:266)
	at java.base/java.util.Objects.checkIndex(Objects.java:359)
	at java.base/java.util.ArrayList.get(ArrayList.java:427)
	at ru.itmo.sd.shell.cli.command.GrepCommand$pattern$2.invoke(GrepCommand.kt:36)
	at ru.itmo.sd.shell.cli.command.GrepCommand$pattern$2.invoke(GrepCommand.kt:32)
	at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
	at ru.itmo.sd.shell.cli.command.GrepCommand.getPattern(GrepCommand.kt:32)
	at ru.itmo.sd.shell.cli.command.GrepCommand.<init>(GrepCommand.kt:39)
	at ru.itmo.sd.shell.parser.CommandParser$parseSimpleCommand$buildCommand$5.invoke(CommandParser.kt:67)
	at ru.itmo.sd.shell.parser.CommandParser$parseSimpleCommand$buildCommand$5.invoke(CommandParser.kt:67)
	at ru.itmo.sd.shell.parser.CommandParser.parseSimpleCommand(CommandParser.kt:77)
	at ru.itmo.sd.shell.parser.CommandParser.parseCliCommand(CommandParser.kt:49)
	at ru.itmo.sd.shell.parser.CommandParser.parse(CommandParser.kt:28)
	at ru.itmo.sd.shell.processor.CommandProcessor.run(CommandProcessor.kt:18)
	at ru.itmo.sd.shell.ShellKt.main(Shell.kt:7)
	at ru.itmo.sd.shell.ShellKt.main(Shell.kt)
```
`-1`

* Не принимает пайп `cat gradlew | grep java`. `-1`.

* Не обрабатываются неправильные ключи:
```
grep -A -1 Java gradlew
____________________
# Determine the Java command to use to start the JVM.
____________________
location of your Java installation."
____________________
location of your Java installation."

grep -asdasd Java gradlew
grep: Java: No such file or directory
Java
```
`-0.5`

* Падает:
```
grep -A java gradlew
Exception in thread "main" java.lang.NumberFormatException: For input string: "java"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
	at java.base/java.lang.Integer.parseInt(Integer.java:660)
	at java.base/java.lang.Integer.parseInt(Integer.java:778)
	at ru.itmo.sd.shell.cli.command.GrepCommand.<init>(GrepCommand.kt:19)
	at ru.itmo.sd.shell.parser.CommandParser$parseSimpleCommand$buildCommand$5.invoke(CommandParser.kt:67)
	at ru.itmo.sd.shell.parser.CommandParser$parseSimpleCommand$buildCommand$5.invoke(CommandParser.kt:67)
	at ru.itmo.sd.shell.parser.CommandParser.parseSimpleCommand(CommandParser.kt:77)
	at ru.itmo.sd.shell.parser.CommandParser.parseCliCommand(CommandParser.kt:49)
	at ru.itmo.sd.shell.parser.CommandParser.parse(CommandParser.kt:28)
	at ru.itmo.sd.shell.processor.CommandProcessor.run(CommandProcessor.kt:18)
	at ru.itmo.sd.shell.ShellKt.main(Shell.kt:7)
	at ru.itmo.sd.shell.ShellKt.main(Shell.kt)
```
`-0.5`

Пожалуйста исправьте замечания.
`6/10`.

ArsenHD committed:
Fix extra newlines in echo command

ArsenHD committed:
Fix variable substitution + brackets processing

ArsenHD committed:
Make commands use bytewise read/write operations

ArsenHD committed:
Write error messages to a separate error stream

ArsenHD committed:
Simplify execution result API

ArsenHD committed:
Fix: remove double quotes from input in all cases

ArsenHD committed:
Use abstract factory to create command objects

This makes parser independent from the set of supported commands

ArsenHD committed:
Modify lexer and parser to support multiline input

ArsenHD committed:
Remove unnecessary property from ExternalCommand

ArsenHD committed:
Make ExecutionFailureError an inheritor of Exception

ArsenHD committed:
Parametrize parser by lexer

ArsenHD committed:
Move command factory up in package hierarchy

ArsenHD committed:
Make CliElement not sealed to allow external users to extend commands hierarhy

ArsenHD committed:
Allow application to finish gracefully upon exit command

ArsenHD committed:
Make empty list a default value for command arguments

ArsenHD committed:
Parametrize commands with IO streams + store piped commands in lists

ArsenHD committed:
Minor formatting fix

ArsenHD committed:
Print errors to output like bash does

ArsenHD committed:
Fix creation of process for external command

ArsenHD committed:
Stop 'wc' command execution when end of transmission character is encountered

ArsenHD committed:
Add IO streams customization to command factories

ArsenHD committed:
Allow all commands to have customized IO streams

ArsenHD committed:
Use executor service to run pipeline commands in threads

ArsenHD committed:
Use executor service to run pipeline commands in threads

ArsenHD committed:
Refactor parser instantiation

ArsenHD committed:
Parse commands with customized IO streams

ArsenHD committed:
Fix main loop error isolation

ArsenHD committed:
Refactor tests: use factories for command instantiation, better arguments, move out common code

ArsenHD committed:
Add parser tests

ArsenHD committed:
Allow input reading to be interrupted when the pipeline is complete, but the input stream is not

ArsenHD committed:
Do not close the given IO streams, but rather the created reader and writer

The IO streams should be handled by the one who created them, not by the command

ArsenHD committed:
Remove unused exceptions

ArsenHD committed:
Add environment to command and let all types of cli elements be executed

ArsenHD committed:
Make test data more clear

ArsenHD committed:
Forbid unmatched quotes

ArsenHD committed:
Add 'private' modifier to a property

ArsenHD committed:
Add .gitignore

ArsenHD committed:
Add .editorconfig for Kotlin code

ArsenHD committed:
Add linter to CI + trigger CI on all branches

ArsenHD committed:
Minor code clean up

ArsenHD committed:
Save test run results in artifacts

ArsenHD committed:
Finish reading remaining bytes after interruption

ArsenHD committed:
Add code coverage computation to CI

ArsenHD committed:
Add more tests on parsing and a test on wc command with files

ArsenHD committed:
Use JaCoCo to compute code coverage

ArsenHD committed:
Make wc test work on linux

ArsenHD committed:
Fix JaCoCo .csv file location

ArsenHD committed:
Fix code coverage badge generation

ArsenHD committed:
Add coverage badge to README.md

ArsenHD committed:
Add fatJar task and upload CI artifacts

ArsenHD committed:
Make README.md more detailed

ArsenHD committed:
Use unix-style line separator in gradlew and run.sh

ArsenHD committed:
Use custom output stream for messages to the user instead of always using stdout

ArsenHD committed:
Print a start message when the interpreted is started

ArsenHD committed:
Hide gradle build output when running shell with a script run.sh

ArsenHD committed:
Add current directory info to environment

ArsenHD committed:
Make program free all its resources when it exits by any reason

ArsenHD committed:
Remove unused dependency

ArsenHD committed:
Handle incorrect options

ArsenHD committed:
Add tests on incorrect options and missing file for grep command

ArsenHD committed:
Minor readme update

ArsenHD committed:
Autogenerated JaCoCo coverage badge

ArsenHD reviewed (commented):
Исправил код по замечаниям и починил функциональность (пайп с grep, обработка неправильных опций, падающий grep)

FirstTimeInForever commented:
Тут все окей. `10/10`.
